lib/common/redis.c

```diff
@@ -0,0 +1,219 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#include "async.h"
+#include "h2o/redis.h"
+#include "h2o/socket.h"
+
+static int attach_loop(redisAsyncContext* ac, h2o_loop_t* loop);
+
+struct st_h2o_redis_conn_t {
+    redisAsyncContext *redis;
+    h2o_redis_callbacks_t cb;
+    void *cb_data;
+};
+
+struct st_h2o_redis_callback_args_t {
+    h2o_redis_command_cb cb;
+    void *data;
+};
+
+void on_command(redisAsyncContext *redis, void *reply, void *privdata)
+{
+    struct st_h2o_redis_callback_args_t *args = (struct st_h2o_redis_callback_args_t *)privdata;
+    if (args != NULL) {
+        args->cb((redisReply *)reply, args->data);
+        free(args);
+    }
+}
+
+void h2o_redis_command(h2o_redis_conn_t *conn, h2o_redis_command_cb cb, void *cb_data, const char *format, ...)
+{
+    if (conn->redis == NULL) {
+        if (conn->cb.on_connection_failure != NULL) {
+            conn->cb.on_connection_failure("redis connection has been closed", conn->cb_data);
+        }
+        return;
+    }
+
+    struct st_h2o_redis_callback_args_t *args = NULL;
+    if (cb != NULL) {
+        args = h2o_mem_alloc(sizeof(struct st_h2o_redis_callback_args_t));
+        args->cb = cb;
+        args->data = cb_data;
+    }
+
+    va_list ap;
+    va_start(ap, format);
+    if (redisvAsyncCommand(conn->redis, on_command, args, format, ap) != REDIS_OK) {
+        /* the case that redisAsyncContext is disconnecting or freeing */
+        /* call the callback immediately with NULL reply */
+        on_command(conn->redis, NULL, args);
+    }
+    va_end(ap);
+}
+
+static void on_redis_connect(const redisAsyncContext *redis, int status)
+{
+    h2o_redis_conn_t *conn = (h2o_redis_conn_t *)redis->data;
+    if (status == REDIS_OK) {
+        if (conn->cb.on_connect != NULL) {
+            conn->cb.on_connect(conn->cb_data);
+        }
+    } else {
+        conn->redis = NULL;
+        if (conn->cb.on_connection_failure != NULL) {
+            conn->cb.on_connection_failure(redis->errstr, conn->cb_data);
+        }
+    }
+}
+
+static void on_redis_disconnect(const redisAsyncContext *redis, int status)
+{
+    h2o_redis_conn_t *conn = (h2o_redis_conn_t *)redis->data;
+    if (status == REDIS_OK) {
+        if (conn->cb.on_disconnect != NULL) {
+            conn->cb.on_disconnect(conn->cb_data);
+        }
+    } else {
+        conn->redis = NULL;
+        if (conn->cb.on_connection_failure != NULL) {
+            conn->cb.on_connection_failure(redis->errstr, conn->cb_data);
+        }
+    }
+}
+
+h2o_redis_conn_t *h2o_redis_connect(h2o_loop_t *loop, const char *host, uint16_t port, h2o_redis_callbacks_t cb, void *cb_data)
```

Comment:
I see several issues regarding the design of the function.

I. The function returns failure using two methods, either by returning NULL or by calling a callback. Such approach complicates things. I'd suggest always using a callback for notifying errors.
Note that memory allocation in H2O must be done by `h2o_mem_alloc` (or similar functions defined in h2o/memory.h) that never returns a failure.

II. The function is related to three callbacks: on_connect, on_connection_failure, on_disconnect. I'd suggest merging them into one or two functions (e.g. on_connect and on_close; in which case the latter function would be called for both connection failure and disconnection, with its argument specifying the reason).

III. The interface is designed to use `void *cb_data` to associate the user data to the redis connection. I'd suggest using structure-extensions instead, much like we do for other structures (e.g. h2o_conn_t, h2o_handler_t). That way, in _some_ applications you would not need to maintain the bidirectional pointer that connects h2o_redis_conn_t and the user data.

IV. The proposed model creates `h2o_redis_conn_t` upon connection, and discards the object upon disconnection. It _might_ be worth considering if we should decouple connection/disconnection and the lifecycle of the object. My understanding is that in most of the use cases we would want to retain a connection to a redis server, and use it from many places. Having a single object that occasionally connects / disconnects to a redis server might help in that respect.

V. You should be careful in calling the callbacks synchronously. For example, the get callback of async resumption is called from a callback within SSL_handshake. And I believe there is a path that leading h2o_socket_ssl_resume_server_handshake before exiting from the get callback. What this means that SSL_handshake (called from h2o_socket_ssl_resume_server_hanshake) would be called within SSL_handshake (that is calling the get callback). However, SSL_handshake is not a reentrant function. I'd suggest to make all callback invocations asynchronous, by using zero_timeout when necessary.